### PR TITLE
fix(compass-components): improve dark theme toggle appearance COMPASS-5470

### PR DESCRIPTION
**Pull request by Anemy**
COMPASS-5470
This PR adds a theme context to the react tree that the connections page is in. This theme context is used in the toggle to render the leafygreen dark theme when dark theme is turned on.

This does not yet fix the toggles in the aggregations view as those are currently in a different react tree and do not have access to the theme context. Work to make them use one react root will happen in COMPASS-5481 . Once that is done, these changes will fix those toggles as well.

I think we could initialize the theme preference, listen to updates in the theme from the main compass process, and propagate that state to the react renderer in a more concise, generalized way which would hopefully free up adding more settings to Compass. I'll leave a comment where this is in the code with this text also. Since the theme is the only state of Compass which is currently managed between multiple windows, I think it's alright to have the solution in this pr for now to unblock the connect form and give the theme to the react context. We can improve and generalize the initialization and synchronization of the shared state in future work when there's more to sync. What do y'all think? Should we do more now?

Before:
Screen Shot 2022-02-02 at 1 44 11 PM

After:
Screen Shot 2022-02-02 at 1 43 48 PM

### CODE REVIEW

File: packages/compass/src/app/index.js
Comment (Anemy) [inline packages/compass/src/app/index.js:75]: I think we could initialize the theme preference, listen to updates in the theme from the main compass process, and propagate that state to the react renderer in a more concise, generalized way which would hopefully free up adding more settings to Compass. Since the theme is the only state of Compass which is currently managed between multiple windows, I think it's alright to have the solution in this pr for now to unblock the connect form and give the theme to the react context. We can improve and generalize the initialization and synchronization of the shared state in future work when there's more to sync. What do y'all think? Should we do more now?

**Comment by Anemy** [comment]
Hmm seeing two different build ci e2e tests failing for timeout after merging main. Could be a rendering slowdown from these changes or unrelated. Looking into

File: packages/compass/src/app/index.js
Comment (mcasimir) [reply packages/compass/src/app/index.js:75]: I think is fine for this PR, we can do that work when we will add user settings.

One improvement I'd really love to see is to extract all of this "theme" related code from this already bloated `index.js`, including the initialization: https://github.com/mongodb-js/compass/blob/dad2a58c23b888f2b832bb6458e155b1f6ccff5c/packages/compass/src/app/index.js#L391

We did something similar for the logging: https://github.com/mongodb-js/compass/blob/ca7f24fca54d7249696821edb1417301253aad3e/packages/compass/src/main/logging.ts

Among just cleaning up that file a bit would also enable us to write some tests for it at some point

**Review by mcasimir — APPROVED** [review]
This is really cool

File: packages/compass-home/src/components/home.tsx
Comment (Anemy) [reply packages/compass-home/src/components/home.tsx:276]: It was because it wasn't being used elsewhere. I'll move it up now that https://github.com/mongodb-js/compass/pull/2754 is landing 👍

Comment (Anemy) [reply packages/compass-home/src/components/home.tsx]: updated

File: packages/compass-components/src/hooks/use-theme.tsx
Comment (Anemy) [reply packages/compass-components/src/hooks/use-theme.tsx]: updated 👍

File: packages/compass-components/src/components/toggle.tsx
Comment (gribnoysup) [inline packages/compass-components/src/components/toggle.tsx:7]:
```suggestion
  props: Omit<React.ComponentProps<typeof LeafyGreenToggle>, 'darkMode'>
```

File: packages/compass-components/src/components/toggle.tsx
Comment (gribnoysup) [inline packages/compass-components/src/components/toggle.tsx:12]: Raised this with leafygreen, if that's the only way to set the darkMode for their components, we are at the risk of wrapping every single leafygreen component that we export here which sounds like a giant maintenance headache

Comment (mcasimir) [reply packages/compass-components/src/components/toggle.tsx:12]: yes it would be massively better if they'd have a provider we can use, I'd expect LG to also provide themes and guidelines for the palette, which is something we'd have to do on our side otherwise:

```
const darkTheme = {
    theme: Theme.Dark
    colors: {
        accent: palette.green.light1,
        // .. etc
    }
}
```

Comment (mcasimir) [reply packages/compass-components/src/components/toggle.tsx:12]: We could do something like this before re-exporting the components on our side:

```
const LGComponentWithTheme<T>(Wrappee: T) {
   const theme = useTheme()
   return (props) => Wrappee(...props, {...darkmode: theme?.theme})
}
```

```
const Toggle = LGComponentWithTheme(Toggle)
export Toggle;
```

Comment (gribnoysup) [reply packages/compass-components/src/components/toggle.tsx:12]: I mean, yeah, we totally can do something like that, HOC is indeed a good pattern to use here. It's just weird that we have even though we all agree that we want to do as little changes to leafygreen as possible

Comment (gribnoysup) [reply packages/compass-components/src/components/toggle.tsx:12]: lg team [said](https://mongodb.slack.com/archives/CFFA5BS79/p1643962820401349) this will be addressed when they are eventually working on a dark mode support

Comment (Anemy) [reply packages/compass-components/src/components/toggle.tsx:12]: What should we do for now? I think we probably won't need to do many fixes like this until dark theme/the new provider is released in leafygreen so having one component listening to the theme is probably alright?

File: packages/compass-components/src/components/toggle.tsx
Comment (Anemy) [reply packages/compass-components/src/components/toggle.tsx:7]: Good question. I think we probably want it to be overrideable, yes, but I'm alright with either approach. If it's a component which always is displayed in a darkMode or always needs to be light mode I think we'll want to customize it and override the provider. I think we could also add more providers in that situation, but if it's just one component I think it's simpler to just override this. I'm cool with either here.

File: packages/compass-components/src/components/toggle.tsx
Comment (mcasimir) [reply packages/compass-components/src/components/toggle.tsx:12]: I would merge this as it is

File: packages/compass-components/src/components/toggle.tsx
Comment (gribnoysup) [reply packages/compass-components/src/components/toggle.tsx:7]: Yeah, this makes sense! Let's keep it like that for now 👍

**Pull request merged**
